File: clients/typescript/test/predicateWrite.test.ts
import assert from "node:assert/strict";
import { after, test } from "node:test";

import { start, type Serving } from "./harness.js";
import {
  add,
  col,
  eq,
  ge,
  lit,
  lt,
  type Session,
  type Value,
  int,
  str,
  uint,
} from "../src/index.js";

/**
 * Predicate writes over the wire, and `returning`.
 *
 * `returning` is offered on these two and on nothing else. Over this wire an
 * insert cannot produce anything a caller does not already have: the proto's
 * row is full width and its nulls are values a caller meant, so there is no
 * unset column for a `DEFAULT` to fill, and there is no auto-increment, no
 * trigger and no generated column. The row written is the row sent. A
 * predicate write is the other case — the caller named a *condition*, and
 * which rows matched is a fact it does not have; for a delete it is a fact
 * that stops existing the moment the write lands.
 */
const PREDICATE_TABLES = `
[[tables]]
name = "papers"
id = 41
columns = [
  { name = "id",   type = "u64" },
  { name = "kind", type = "str" },
  { name = "size", type = "i64" },
]
primary_key = ["id"]

[[security.grants]]
role = "app"
tables = ["papers"]
actions = ["everything"]
`;

const ID = 0;
const KIND = 1;
const SIZE = 2;
const COUNT = 12;

const servers: Serving[] = [];
after(() => {
  for (const s of servers) s.stop();
});

/** A node holding twelve papers: ids 0..11, size = id, kind-0..kind-3. */
async function seeded(): Promise<Session> {
  const server = await start(PREDICATE_TABLES);
  servers.push(server);
  const session = server.client().session();
  const rows: Value[][] = [];
  for (let id = 0; id < COUNT; id++) {
    rows.push([uint(id), str(`kind-${id % 4}`), int(id)]);
  }
  await session.insert("papers", ...rows);
  return session;
}

async function remaining(session: Session): Promise<number> {
  let seen = 0;
  for await (const _ of session.query({ table: "papers" })) seen++;
  return seen;
}

function ids(rows: Value[][]): bigint[] {
  return rows.map((row) => {
    const id = row[0];
    assert.equal(id?.kind, "uint", `the first column was ${id?.kind}`);
    return (id as { value: bigint }).value;
  });
}

function sizes(rows: Value[][]): bigint[] {
  return rows.map((row) => {
    const size = row[2];
    assert.equal(size?.kind, "int", `size was ${size?.kind}`);
    return (size as { value: bigint }).value;
  });
}

test("a predicate delete removes every row it selects", async () => {
  const session = await seeded();
  const result = await session.deleteWhere({
    table: "papers",
    filter: ge(SIZE, int(9)),
  });
  assert.equal(result.affected, 3n);
  assert.deepEqual(result.rows, [], "no rows were asked for");
  assert.equal(await remaining(session), 9);
});

test("a predicate delete returns the rows it destroyed", async () => {
  // The one that could not be written any other way: after the delete the rows
  // are gone, so this response is the only record of what they were.
  const session = await seeded();
  const result = await session.deleteWhere({
    table: "papers",
    filter: ge(SIZE, int(9)),
    returning: true,
  });
  assert.equal(result.affected, 3n);
  assert.deepEqual(ids(result.rows), [9n, 10n, 11n]);
  assert.deepEqual(sizes(result.rows), [9n, 10n, 11n]);
  assert.equal(await remaining(session), 9);
});

test("a predicate update returns the rows as written", async () => {
  const session = await seeded();
  const result = await session.updateWhere({
    table: "papers",
    filter: lt(SIZE, int(3)),
    // size = size + 100: one write, not a read, a decision and a write.
    set: [{ column: SIZE, value: add(col(SIZE), lit(int(100))) }],
    returning: true,
  });
  assert.equal(result.affected, 3n);
  assert.deepEqual(ids(result.rows), [0n, 1n, 2n]);
  assert.deepEqual(
    sizes(result.rows),
    [100n, 101n, 102n],
    "the rows come back as written, not as they were",
  );
});

test("every assignment reads the original row", async () => {
  const session = await seeded();
  const result = await session.updateWhere({
    table: "papers",
    filter: eq(ID, uint(3)),
    set: [
      { column: SIZE, value: add(col(SIZE), lit(int(1))) },
      { column: KIND, value: col(KIND) },
    ],
    returning: true,
  });
  assert.equal(result.affected, 1n);
  assert.deepEqual(sizes(result.rows), [4n]);
});

test("an absent filter is every row", async () => {
  const session = await seeded();
  const result = await session.deleteWhere({ table: "papers" });
  assert.equal(result.affected, BigInt(COUNT));
  assert.equal(await remaining(session), 0);
});

test("a predicate that matches nothing writes nothing", async () => {
  const session = await seeded();
  const result = await session.deleteWhere({
    table: "papers",
    filter: eq(KIND, str("no-such-kind")),
    returning: true,
  });
  assert.equal(result.affected, 0n);
  assert.deepEqual(result.rows, []);
  assert.equal(await remaining(session), COUNT);
});

test("an update with no assignments is refused", async () => {
  // Refused rather than reported as zero rows written: zero is what a
  // predicate that matched nothing reports, and the two are different
  // mistakes.
  const session = await seeded();
  await assert.rejects(
    () => session.updateWhere({ table: "papers", set: [] }),
    (error: Error) => {
      assert.match(error.message, /assignment/);
      return true;
    },
  );
  assert.equal(await remaining(session), COUNT);
});

test("a column assigned twice is refused", async () => {
  const session = await seeded();
  await assert.rejects(() =>
    session.updateWhere({
      table: "papers",
      set: [
        { column: SIZE, value: lit(int(1)) },
        { column: SIZE, value: lit(int(2)) },
      ],
    }),
  );
  assert.equal(await remaining(session), COUNT);
});

File: clients/go/slate/predicate_write_test.go
package slate_test

import (
	"strings"
	"testing"

	"github.com/howlerops/slate-orm/clients/go/slate"
)

const predicateTables = `
[[tables]]
name = "papers"
id = 40
columns = [
  { name = "id",   type = "u64" },
  { name = "kind", type = "str" },
  { name = "size", type = "i64" },
]
primary_key = ["id"]

[[security.grants]]
role = "app"
tables = ["papers"]
actions = ["everything"]
`

const (
	paperID    slate.Ordinal = 0
	paperKind  slate.Ordinal = 1
	paperSize  slate.Ordinal = 2
	paperCount               = 12
)

// seeded starts a node holding twelve papers: ids 0..11, size = id, and
// kind-0..kind-3 round-robin.
func seeded(t *testing.T) *slate.Session {
	t.Helper()
	session := start(t, predicateTables).client(t).Session()
	ctx := testContext(t)
	rows := make([][]slate.Value, 0, paperCount)
	for id := uint64(0); id < paperCount; id++ {
		rows = append(rows, []slate.Value{
			slate.Uint(id),
			slate.String("kind-" + string(rune('0'+id%4))),
			slate.Int(int64(id)),
		})
	}
	if _, err := session.Insert(ctx, "papers", rows...); err != nil {
		t.Fatalf("seeding: %v", err)
	}
	return session
}

// remaining is every id still in the table.
func remaining(t *testing.T, session *slate.Session) []uint64 {
	t.Helper()
	stream, err := session.Query(testContext(t), slate.Query{Table: "papers"})
	if err != nil {
		t.Fatalf("querying: %v", err)
	}
	defer stream.Close()
	out := make([]uint64, 0, paperCount)
	// Row() advances the cursor; Next() alone only reports that one is there,
	// so a loop that never reads spins forever. The idiom is load-bearing.
	for stream.Next() {
		row := stream.Row()
		id, ok := row[0].(slate.Uint)
		if !ok {
			t.Fatalf("the first column was not a u64: %v", row[0])
		}
		out = append(out, uint64(id))
	}
	if err := stream.Err(); err != nil {
		t.Fatalf("stream: %v", err)
	}
	return out
}

func returnedIDs(t *testing.T, result slate.WriteResult) []uint64 {
	t.Helper()
	out := make([]uint64, 0, len(result.Rows))
	for _, row := range result.Rows {
		id, ok := row[0].(slate.Uint)
		if !ok {
			t.Fatalf("the first column was not a u64: %v", row[0])
		}
		out = append(out, uint64(id))
	}
	return out
}

func TestDeleteWhereRemovesEveryMatchingRow(t *testing.T) {
	session := seeded(t)
	result, err := session.DeleteWhere(testContext(t), slate.DeleteWhere{
		Table: "papers", Filter: slate.Filter(slate.Ge(paperSize, slate.Int(9))),
	})
	if err != nil {
		t.Fatalf("deleting: %v", err)
	}
	if result.Affected != 3 {
		t.Fatalf("affected = %d, want 3", result.Affected)
	}
	if len(result.Rows) != 0 {
		t.Fatalf("no rows were asked for, got %d", len(result.Rows))
	}
	if got := len(remaining(t, session)); got != 9 {
		t.Fatalf("%d rows left, want 9", got)
	}
}

// TestDeleteWhereReturnsWhatItDestroyed is the one that could not be written
// any other way: after the delete the rows are gone, so this response is the
// only record of what they were.
func TestDeleteWhereReturnsWhatItDestroyed(t *testing.T) {
	session := seeded(t)
	result, err := session.DeleteWhere(testContext(t), slate.DeleteWhere{
		Table:     "papers",
		Filter:    slate.Filter(slate.Ge(paperSize, slate.Int(9))),
		Returning: true,
	})
	if err != nil {
		t.Fatalf("deleting: %v", err)
	}
	got := returnedIDs(t, result)
	if len(got) != 3 || got[0] != 9 || got[1] != 10 || got[2] != 11 {
		t.Fatalf("returned %v, want [9 10 11]", got)
	}
	// The sizes came back too, which a caller holding only a predicate could
	// not have reconstructed once the rows were gone.
	for at, row := range result.Rows {
		size, ok := row[2].(slate.Int)
		if !ok || int64(size) != int64(9+at) {
			t.Fatalf("row %d size = %v, want %d", at, row[2], 9+at)
		}
	}
}

func TestUpdateWhereReturnsRowsAsWritten(t *testing.T) {
	session := seeded(t)
	result, err := session.UpdateWhere(testContext(t), slate.UpdateWhere{
		Table:  "papers",
		Filter: slate.Filter(slate.Lt(paperSize, slate.Int(3))),
		// size = size + 100: one write, not a read, a decision and a write.
		Set:       []slate.Assignment{slate.Assign(paperSize, slate.Add(slate.Col(paperSize), slate.Lit(slate.Int(100))))},
		Returning: true,
	})
	if err != nil {
		t.Fatalf("updating: %v", err)
	}
	if result.Affected != 3 {
		t.Fatalf("affected = %d, want 3", result.Affected)
	}
	for at, row := range result.Rows {
		size, ok := row[2].(slate.Int)
		if !ok || int64(size) != int64(100+at) {
			t.Fatalf("row %d size = %v, want %d — the rows come back as written", at, row[2], 100+at)
		}
	}
}

// TestUpdateWhereAppliesAssignmentsTogether pins that every assignment reads
// the original row, so one cannot see another's result.
func TestUpdateWhereAppliesAssignmentsTogether(t *testing.T) {
	session := seeded(t)
	result, err := session.UpdateWhere(testContext(t), slate.UpdateWhere{
		Table:  "papers",
		Filter: slate.Filter(slate.Eq(paperID, slate.Uint(3))),
		Set: []slate.Assignment{
			slate.Assign(paperSize, slate.Add(slate.Col(paperSize), slate.Lit(slate.Int(1)))),
			slate.Assign(paperKind, slate.Col(paperKind)),
		},
		Returning: true,
	})
	if err != nil {
		t.Fatalf("updating: %v", err)
	}
	if size, ok := result.Rows[0][2].(slate.Int); !ok || int64(size) != 4 {
		t.Fatalf("size = %v, want 4", result.Rows[0][2])
	}
}

// TestDeleteWhereWithNoFilterIsEveryRow: a nil filter is `DELETE FROM docs`,
// which is a real statement and is allowed.
func TestDeleteWhereWithNoFilterIsEveryRow(t *testing.T) {
	session := seeded(t)
	result, err := session.DeleteWhere(testContext(t), slate.DeleteWhere{Table: "papers"})
	if err != nil {
		t.Fatalf("deleting: %v", err)
	}
	if result.Affected != paperCount {
		t.Fatalf("affected = %d, want %d", result.Affected, paperCount)
	}
	if got := remaining(t, session); len(got) != 0 {
		t.Fatalf("%d rows left, want none", len(got))
	}
}

func TestUpdateWhereWithNoAssignmentsIsRefused(t *testing.T) {
	session := seeded(t)
	_, err := session.UpdateWhere(testContext(t), slate.UpdateWhere{Table: "papers"})
	if err == nil {
		t.Fatal("an update with no assignments should be refused")
	}
	if !strings.Contains(err.Error(), "assignment") {
		t.Fatalf("the refusal should say what is missing: %v", err)
	}
	if got := len(remaining(t, session)); got != paperCount {
		t.Fatalf("%d rows left, want %d", got, paperCount)
	}
}

func TestUpdateWhereRefusesAColumnAssignedTwice(t *testing.T) {
	session := seeded(t)
	_, err := session.UpdateWhere(testContext(t), slate.UpdateWhere{
		Table: "papers",
		Set: []slate.Assignment{
			slate.Assign(paperSize, slate.Lit(slate.Int(1))),
			slate.Assign(paperSize, slate.Lit(slate.Int(2))),
		},
	})
	if err == nil {
		t.Fatal("a column assigned twice should be refused")
	}
}

func TestPredicateWriteMatchingNothingWritesNothing(t *testing.T) {
	session := seeded(t)
	result, err := session.DeleteWhere(testContext(t), slate.DeleteWhere{
		Table:     "papers",
		Filter:    slate.Filter(slate.Eq(paperKind, slate.String("no-such-kind"))),
		Returning: true,
	})
	if err != nil {
		t.Fatalf("deleting: %v", err)
	}
	if result.Affected != 0 || len(result.Rows) != 0 {
		t.Fatalf("affected = %d with %d rows, want 0 and none", result.Affected, len(result.Rows))
	}
	if got := len(remaining(t, session)); got != paperCount {
		t.Fatalf("%d rows left, want %d", got, paperCount)
	}
}
